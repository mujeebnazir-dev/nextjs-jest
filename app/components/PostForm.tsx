"use client";

import { useState } from "react";

interface PostFormProps {
  onSubmit: (title: string, content: string) => void;
}

export default function PostForm({ onSubmit }: PostFormProps) {
  const [title, setTitle] = useState("");
  const [content, setContent] = useState("");
  const [isSubmitting, setIsSubmitting] = useState(false);

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    
    if (!title.trim() || !content.trim()) {
      return ;
    }

    setIsSubmitting(true);
    
    try {
      onSubmit(title.trim(), content.trim());
      setTitle("");
      setContent("");
    } catch (error) {
      console.error("Error submitting post:", error);
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <div className="bg-white rounded-lg shadow-sm p-6 border">
      <h2 className="text-xl font-semibold text-gray-800 mb-4">Create New Post</h2>
      
      <form onSubmit={handleSubmit} className="space-y-4">
        <div>
          <label htmlFor="title" className="block text-sm font-medium text-gray-700 mb-1">
            Title
          </label>
          <input
            id="title"
            type="text"
            placeholder="Enter post title..."
            className="w-full border border-gray-300 rounded-lg px-3 text-black py-2 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
            value={title}
            onChange={(e) => setTitle(e.target.value)}
            disabled={isSubmitting}
            maxLength={100}
          />
          <div className="text-right text-xs text-gray-500 mt-1">
            {title.length}/100
          </div>
        </div>

        <div>
          <label htmlFor="content" className="block text-sm font-medium text-gray-700 mb-1">
            Content
          </label>
          <textarea
            id="content"
            placeholder="Write your post content here..."
            className="w-full border border-gray-300 rounded-lg px-3 py-2 text-black focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 resize-vertical"
            rows={4}
            value={content}
            onChange={(e) => setContent(e.target.value)}
            disabled={isSubmitting}
            maxLength={1000}
          />
          <div className="text-right text-xs text-gray-500 mt-1">
            {content.length}/1000
          </div>
        </div>

        <div className="flex justify-end">
          <button
            type="submit"
            className="px-6 py-2 bg-blue-600 text-white rounded-lg hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300 disabled:cursor-not-allowed transition-colors"
            disabled={isSubmitting || !title.trim() || !content.trim()}
          >
            {isSubmitting ? "Publishing..." : "Publish Post"}
          </button>
        </div>
      </form>
    </div>
  );
}