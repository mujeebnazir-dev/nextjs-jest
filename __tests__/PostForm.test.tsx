import { render, screen, fireEvent } from "@testing-library/react";
import PostForm from "../app/components/PostForm";
import { expect, jest, test, describe } from "@jest/globals";

describe("PostForm - Test suit using Jest", () => {

  /* Snapshot testing */
  test("matches snapshot", () => {
  const { asFragment } = render(<PostForm onSubmit={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});

  /* Test case 1: check if form fields and button render correctly */
  test("renders form fields and button", () => {
    render(<PostForm onSubmit={() => {}} />);
    expect(screen.getByLabelText(/Title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Content/i)).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Publish Post/i })
    ).toBeInTheDocument();
  });

  /* Test case 2: submit button disabled when fields are empty */
  test("disables submit button when fields are empty", () => {
    render(<PostForm onSubmit={() => {}} />);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  /* Test case 3: calls onSubmit with trimmed values when submitted */
  test("calls onSubmit with trimmed values when submitted", () => {
    const mockSubmit = jest.fn();
    render(<PostForm onSubmit={mockSubmit} />);

    fireEvent.change(screen.getByLabelText(/Title/i), {
      target: { value: "  Hello " },
    });
    fireEvent.change(screen.getByLabelText(/Content/i), {
      target: { value: "  World " },
    });
    fireEvent.click(screen.getByRole("button"));

    expect(mockSubmit).toHaveBeenCalledTimes(1);
    expect(mockSubmit).toHaveBeenCalledWith("Hello", "World");
  });

  /* Test case 4: fields are cleared after submission */
  test("clears fields after submit", () => {
    const mockSubmit = jest.fn();
    render(<PostForm onSubmit={mockSubmit} />);

    fireEvent.change(screen.getByLabelText(/Title/i), {
      target: { value: "Test" },
    });
    fireEvent.change(screen.getByLabelText(/Content/i), {
      target: { value: "Content" },
    });
    fireEvent.click(screen.getByRole("button"));

    expect(screen.getByLabelText(/Title/i)).toHaveValue("");
    expect(screen.getByLabelText(/Content/i)).toHaveValue("");
  });

  
  /* Test case 5: logs error if onSubmit throws */
  test("logs error if onSubmit throws", () => {
    const error = new Error("submit failed");
    const mockSubmit = jest.fn(() => { throw error; });
    const consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {});

    render(<PostForm onSubmit={mockSubmit} />);
    fireEvent.change(screen.getByLabelText(/Title/i), { target: { value: "Test" } });
    fireEvent.change(screen.getByLabelText(/Content/i), { target: { value: "Content" } });
    fireEvent.click(screen.getByRole("button"));

    expect(consoleErrorSpy).toHaveBeenCalledWith("Error submitting post:", error);

    consoleErrorSpy.mockRestore();
  });
});


