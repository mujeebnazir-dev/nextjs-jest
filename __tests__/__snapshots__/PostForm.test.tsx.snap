// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PostForm - Test suit using Jest matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white rounded-lg shadow-sm p-6 border"
  >
    <h2
      class="text-xl font-semibold text-gray-800 mb-4"
    >
      Create New Post
    </h2>
    <form
      class="space-y-4"
    >
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="title"
        >
          Title
        </label>
        <input
          class="w-full border border-gray-300 rounded-lg px-3 text-black py-2 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          id="title"
          maxlength="100"
          placeholder="Enter post title..."
          type="text"
          value=""
        />
        <div
          class="text-right text-xs text-gray-500 mt-1"
        >
          0/100
        </div>
      </div>
      <div>
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="content"
        >
          Content
        </label>
        <textarea
          class="w-full border border-gray-300 rounded-lg px-3 py-2 text-black focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 resize-vertical"
          id="content"
          maxlength="1000"
          placeholder="Write your post content here..."
          rows="4"
        />
        <div
          class="text-right text-xs text-gray-500 mt-1"
        >
          0/1000
        </div>
      </div>
      <div
        class="flex justify-end"
      >
        <button
          class="px-6 py-2 bg-blue-600 text-white rounded-lg hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300 disabled:cursor-not-allowed transition-colors"
          disabled=""
          type="submit"
        >
          Publish Post
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
